docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+  to find the latest index_part file.
+
+The pageserver must persist this attachment generation number before starting
+the `Tenant` tasks. We will create a new per-tenant metadata file that records
+the attachment generation number, and will also be used in future work for
+implementing fast migration, where the tenant has more states than simple
+attached & detached.
+
+#### Reconciliation
+
+On attachment, the `Tenant` may have some local disk content from previous
+activity involving this tenant (e.g. it used to be attached here), and this
+needs reconciling with remote storage. Note that "reconcile" in this context is
+used generically, and does not refer specifically to the logic currently
+in the `reconcile_with_remote` function.
+
+Because pageservers have access to the WAL, they don't strictly have to
+use the latest remote storage contents at all: if the WAL history still contains
+the data required to bring the pageserver's local state up to date with
+the tip of the WAL, the pageserver could recover from the WAL and then
+overwrite whatever is in remote storage.
+
+However, it is preferable to recover from remote storage rather than
+to replay WAL, for the following reasons:
+
+- Simplicity: recovering from remote storage gives identical behavior
+  whether attaching a tenant for the first time or re-attaching later.
+- Efficiency: ingesting the WAL has significant cost beyond the I/O
+  of reading it (compaction, image layer generation). We should avoid
+  doing this work twice if there is already remote content available.
+- Reduced load on safekeepers, which are a less scalable resource than S3.
+
+#### Finding the remote indices for timelines
+
+Because index files are now suffixed with generation numbers, the pageserver
+cannot always GET the remote index in one request, because it can't always
+know a-priori what the latest remote index is.
+
+In the general case and as a fallback, the pageserver may list all the index*part.json
+files for a timeline, sort them by generation suffix, and pick the highest that is <=
+the suffix for its current generation numbers. The tenant should never load an index with
+an attachment generation \_newer* than its own: tenants
+are allowed to be attached with stale attachment generations during a multiply-attached
+phase in a migration, and in this instance if the old location's pageserver restarts,
+it should not try and load the newer generation's index.
+
+To avoid object listing in most cases, a couple of optimizations can be used:
+
+- On a restart of a previously attached tenant, try subtracting one from our
+  node generation number: if we find an index_part.json with that suffix, it
+  should be the last one we wrote out before restarting.
+- Otherwise, consult the previous attachment information that was provided by
+  the control plane in [attachment](#attachment), and try loading with that
+  suffix.
+- If neither of the above return a 200, then fall back to doing an object listing.
```

Comment:
Alternative:

When a pageserver uploads the `index_part.json_<generation>` file to S3 for the first time after being attached, it tells the control plane that the upload was performed. Control plane remembers that, and when a new attachment is made, it passes it to the new attached pageserver.

In other words, we can arrange things so that the optimization 2 here always succeeds, and we never need to fall back to object listing.

Reply:
Because notifying the control plane and doing the upload can fail independently, this doesn't provide a guarantee: the pageserver might crash between uploading the index and telling the control plane about it, or if we order in the other direction then we might crash between telling the control plane about it and uploading it.  Ot the control plane might crash after issuing a generation number, and thereby be unavailable to receive the notification.

We could implement a notification anyway to improve the chances that the optimization 2 succeeds.  Since we already call into the control plane to validate generations before deletions, that could be taken as a hint by the control plane that the current (attach_gen, node_gen) tuple has just written an index_part.json.